### PR TITLE
Elasticsearch

### PR DESCRIPTION
**Pull request by pkarman**
Basic infrastructure for building ES indices via rake task. One API endpoint partially implemented for Story model.

Still TODO via Epic #444
* [ ] Re-implement API filters against ES index for Story #441
* [ ] Re-implement Networks search for non-Story models #442
* [ ] Incremental updates #443

### CODE REVIEW

File: app/models/concerns/searchable.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    def remove_from_index
+      __elasticsearch__.destroy_document
+      refresh_index
+    end
+
+    def refresh_index
+      self.class.__elasticsearch__.refresh_index!
+    end
+
+    def self.rebuild_index
```
Comment (cavis) [inline app/models/concerns/searchable.rb]: What is this used for?  (Not called anywhere in this PR).  Is this the high-availability async reindexing?

File: app/services/story_query_builder.rb [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    if clauses.length == 2
+      clauses.map { |c| "(#{c})" }.join(" AND ")
+    elsif clauses.length == 1
+      clauses[0].to_s
+    else
+      ""
+    end
+  end
+
+  def munge_fielded_params(fielded)
```
Comment (cavis) [inline app/services/story_query_builder.rb]: What is this used for?  Some sort of created-in-range query that we don't use yet?

File: app/services/story_query_builder.rb [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  end
+
+  def published_filter
+    searchdsl = self
+    Filter.new do
+      term published: true
+    end
+  end
+
+  def authz_filter
```
Comment (cavis) [inline app/services/story_query_builder.rb]: Guessing that a lot of these methods will eventually be shared with other types of query too.  (all the other models where you `include Searchable`)

File: bin/application [hunk 4]
```diff
@@ -29,6 +29,20 @@ WaitForMysql () {
   done
 }
 
+WaitForElasticsearch () {
```
Comment (cavis) [inline bin/application]: I see how this makes sense for the tests.  But I wonder if ES-availability is really necessary for the CMS app to run at all.  (I'm always hesitant to add app-startup dependencies)

Comment (cavis) [reply bin/application]: Even in dev, I can see it being handy to be able to start CMS without ES.  The search endpoints just won't work.

File: config/routes.rb [hunk 5]
```diff
@@ -16,6 +16,7 @@
 
       resources :stories, except: [:new, :edit] do
         get 'random', on: :collection
+        get 'search', on: :collection
```
Comment (cavis) [inline config/routes.rb]: Hrmm... yeah, remind me what this is for, if we're just overriding the `?text` filter.

File: docker-compose.yml [hunk 6]
```diff
@@ -57,3 +59,18 @@ pg:
     - "5432"
   volumes:
     - ./test/db/:/docker-entrypoint-initdb.d/
+elasticsearch:
+  image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.3.0
+  container_name: elasticsearch
+  environment:
+    - cluster.name=docker-cluster
+    - bootstrap.memory_lock=true
+    - "ES_JAVA_OPTS=-Xms512m -Xmx512m"
```
Comment (cavis) [inline docker-compose.yml]: Hmm... maybe these other ENVs are no longer necessary, if this changes to single-node mode.  (And should change this one to match the docker-compose-ci)

File: app/controllers/api/stories_controller.rb [hunk 7]
```diff
@@ -94,7 +99,9 @@ def resources_base
 
   def filtered(resources)
     resources = resources.v4 if filters.v4?
-    resources = resources.match_text(filters.text) if filters.text?
+    Rails.logger.debug("before: #{resources}")
```
[the remark this answers is not in the document]
Comment (pkarman) [reply app/controllers/api/stories_controller.rb]: no. I'll zap.

File: app/services/story_query_builder.rb (hunk 2 above)
Comment (pkarman) [reply app/services/story_query_builder.rb]: yes. I basically lifted a query builder class I built a couple years ago and started adapting. There are several vestigial pieces. I will mark those so it's clear what isn't (yet) used.

File: docker-compose.yml (hunk 6 above)
Comment (pkarman) [reply docker-compose.yml]: yeah I think I cargo-culted most of the env stuff from an ES example.

**Comment by pkarman** [comment]
@cavis I commented and reverted as much as I could find, and added tests for the new ES functionality.

Since this PR just implements the basic infrastructure and I didn't want it to get too big, I created a number of other tickets to describe what I think are the TODOs. See #444 

I will try adding some unit tests for the DSL query builder. We can wait on those for this PR, or merge in a subsequent PR.

File: bin/application [hunk 8]
```diff
@@ -70,6 +85,7 @@ ApplicationRun () {
 
   UpdateEnvironment
   WaitForMysql
+  WaitForElasticsearch
```
Comment (cavis) [inline bin/application]: (I think maybe you meant to take this out)

File: config/routes.rb (hunk 5 above)
Comment (cavis) [reply config/routes.rb]: +1 on separate endpoint, at least to start.

**Pull request merged**